File: adapters/executor/ollama.go
package executor

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
	"github.com/openai/openai-go/shared"
	"github.com/janmarkuslanger/stagehand/ports"
)

const ollamaDefaultModel = "qwen2.5"
const ollamaDefaultHost = "http://localhost:11434"

// OllamaExecutor runs a task by calling a local Ollama instance via its
// OpenAI-compatible chat completions endpoint.
type OllamaExecutor struct {
	client  *openai.Client
	storage ports.ArtifactStorage
}

// NewOllamaExecutor creates an OllamaExecutor that calls the given host and
// stores artifacts in the provided storage.
func NewOllamaExecutor(host string, storage ports.ArtifactStorage) *OllamaExecutor {
	client := openai.NewClient(
		option.WithBaseURL(host+"/v1"),
		option.WithAPIKey("ollama"),
	)
	return &OllamaExecutor{client: &client, storage: storage}
}

// Execute runs the task described by request against Ollama and returns the
// agent's final text output and any files written during execution.
func (e *OllamaExecutor) Execute(ctx context.Context, request ports.ExecutionRequest) (ports.ExecutionResult, error) {
	model := request.Model
	if model == "" {
		model = ollamaDefaultModel
	}

	tools := buildOllamaTools(request.Tools)

	messages := []openai.ChatCompletionMessageParamUnion{
		openai.UserMessage(request.Prompt),
	}

	if request.SystemPrompt != "" {
		messages = append([]openai.ChatCompletionMessageParamUnion{
			openai.SystemMessage(request.SystemPrompt),
		}, messages...)
	}

	var finalOutput string
	var writtenFiles []string
	var lastFinishReason string

	for step := 0; step < maxAgentSteps; step++ {
		params := openai.ChatCompletionNewParams{
			Model:    model,
			Messages: messages,
		}
		if len(tools) > 0 {
			params.Tools = tools
		}

		resp, err := e.client.Chat.Completions.New(ctx, params)
		if err != nil {
			return ports.ExecutionResult{}, fmt.Errorf("ollama executor: task %s: step %d: %w", request.TaskID, step, err)
		}
		if len(resp.Choices) == 0 {
			return ports.ExecutionResult{}, fmt.Errorf("ollama executor: task %s: step %d: empty response", request.TaskID, step)
		}

		choice := resp.Choices[0]
		lastFinishReason = string(choice.FinishReason)

		if choice.Message.Content != "" {
			finalOutput = choice.Message.Content
		}

		if choice.FinishReason != "tool_calls" {
			break
		}

		messages = append(messages, choice.Message.ToParam())

		var toolResults []openai.ChatCompletionMessageParamUnion
		for _, toolCall := range choice.Message.ToolCalls {
			resultContent, toolErr := e.dispatchOllamaTool(ctx, request.TaskID, toolCall, &writtenFiles)
			if toolErr != nil {
				toolResults = append(toolResults, openai.ToolMessage(toolErr.Error(), toolCall.ID))
			} else {
				toolResults = append(toolResults, openai.ToolMessage(resultContent, toolCall.ID))
			}
		}
		messages = append(messages, toolResults...)
	}

	if lastFinishReason == "tool_calls" {
		return ports.ExecutionResult{}, fmt.Errorf("ollama executor: task %s: agent did not complete within %d steps", request.TaskID, maxAgentSteps)
	}

	return ports.ExecutionResult{Output: finalOutput, Files: writtenFiles}, nil
}

func (e *OllamaExecutor) dispatchOllamaTool(ctx context.Context, taskID string, toolCall openai.ChatCompletionMessageToolCall, writtenFiles *[]string) (string, error) {
	rawInput := json.RawMessage(toolCall.Function.Arguments)
	switch toolCall.Function.Name {
	case "write_file":
		return e.executeOllamaWriteFile(ctx, taskID, rawInput, writtenFiles)
	case "read_file":
		return e.executeOllamaReadFile(ctx, taskID, rawInput)
	case "list_files":
		return e.executeOllamaListFiles(ctx, taskID, rawInput)
	default:
		return "", fmt.Errorf("unknown tool %q", toolCall.Function.Name)
	}
}

func (e *OllamaExecutor) executeOllamaWriteFile(ctx context.Context, taskID string, rawInput json.RawMessage, writtenFiles *[]string) (string, error) {
	var input writeFileInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("write_file: invalid input: %w", err)
	}
	if input.Path == "" {
		return "", fmt.Errorf("write_file: path is required")
	}
	storagePath := taskID + "/" + input.Path
	if err := e.storage.Write(ctx, storagePath, []byte(input.Content)); err != nil {
		return "", fmt.Errorf("write_file: %w", err)
	}
	*writtenFiles = append(*writtenFiles, storagePath)
	return "ok", nil
}

func (e *OllamaExecutor) executeOllamaReadFile(ctx context.Context, taskID string, rawInput json.RawMessage) (string, error) {
	var input readFileInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("read_file: invalid input: %w", err)
	}
	if input.Path == "" {
		return "", fmt.Errorf("read_file: path is required")
	}
	data, err := e.storage.Read(ctx, taskID+"/"+input.Path)
	if err != nil {
		return "", fmt.Errorf("read_file: %w", err)
	}
	return string(data), nil
}

func (e *OllamaExecutor) executeOllamaListFiles(ctx context.Context, taskID string, rawInput json.RawMessage) (string, error) {
	var input listFilesInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("list_files: invalid input: %w", err)
	}
	pattern := taskID + "/*"
	if input.Pattern != "" {
		pattern = taskID + "/" + input.Pattern
	}
	files, err := e.storage.List(ctx, pattern)
	if err != nil {
		return "", fmt.Errorf("list_files: %w", err)
	}
	return strings.Join(files, "\n"), nil
}

func buildOllamaTools(toolNames []string) []openai.ChatCompletionToolParam {
	available := map[string]openai.ChatCompletionToolParam{
		"write_file": {
			Function: shared.FunctionDefinitionParam{
				Name:        "write_file",
				Description: openai.String("Write text content to a file. Creates the file if it does not exist."),
				Parameters: shared.FunctionParameters{
					"type": "object",
					"properties": map[string]any{
						"path":    map[string]any{"type": "string", "description": "Relative file path"},
						"content": map[string]any{"type": "string", "description": "Text content to write"},
					},
					"required": []string{"path", "content"},
				},
			},
		},
		"read_file": {
			Function: shared.FunctionDefinitionParam{
				Name:        "read_file",
				Description: openai.String("Read the text content of a file."),
				Parameters: shared.FunctionParameters{
					"type": "object",
					"properties": map[string]any{
						"path": map[string]any{"type": "string", "description": "Relative file path"},
					},
					"required": []string{"path"},
				},
			},
		},
		"list_files": {
			Function: shared.FunctionDefinitionParam{
				Name:        "list_files",
				Description: openai.String("List files matching a glob pattern."),
				Parameters: shared.FunctionParameters{
					"type": "object",
					"properties": map[string]any{
						"pattern": map[string]any{"type": "string", "description": "Glob pattern, e.g. *.md"},
					},
					"required": []string{"pattern"},
				},
			},
		},
	}

	var result []openai.ChatCompletionToolParam
	for _, name := range toolNames {
		if tool, ok := available[name]; ok {
			result = append(result, tool)
		}
	}
	return result
}

// Compile-time check that OllamaExecutor satisfies ports.AgentExecutor.
var _ ports.AgentExecutor = (*OllamaExecutor)(nil)

File: stagehand/adapters/executor/ollama.py
from __future__ import annotations

import json
from typing import Any, Optional

import openai

from stagehand.adapters.executor.base import (
    BaseAgentExecutor,
    ParsedTurn,
    ToolInvocation,
)
from stagehand.ports.executor import ExecutionRequest, ToolDefinition
from stagehand.ports.storage import ArtifactStorage

OLLAMA_DEFAULT_MODEL = "qwen2.5"
OLLAMA_DEFAULT_HOST = "http://localhost:11434"


class OllamaExecutor(BaseAgentExecutor):
    """Runs a task by calling a local Ollama instance via its OpenAI-compatible endpoint."""

    _label = "ollama executor"

    def __init__(
        self,
        host: str = OLLAMA_DEFAULT_HOST,
        storage: Optional[ArtifactStorage] = None,
        extra_tools: Optional[list[ToolDefinition]] = None,
    ) -> None:
        super().__init__(storage=storage, extra_tools=extra_tools)
        self.client = openai.AsyncOpenAI(
            base_url=f"{host}/v1",
            api_key="ollama",
        )

    def _default_model(self) -> str:
        return OLLAMA_DEFAULT_MODEL

    def _init_messages(self, request: ExecutionRequest) -> list[Any]:
        messages: list[dict[str, Any]] = []
        if request.system_prompt:
            messages.append({"role": "system", "content": request.system_prompt})
        messages.append({"role": "user", "content": request.prompt})
        return messages

    def _serialize_tools(self, tools: list[ToolDefinition]) -> list[dict[str, Any]]:
        return [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.input_schema,
                },
            }
            for t in tools
        ]

    async def _call_model(
        self,
        model: str,
        request: ExecutionRequest,
        messages: list[Any],
        tools: list[dict[str, Any]],
    ) -> Any:
        params: dict[str, Any] = {"model": model, "messages": messages}
        if tools:
            params["tools"] = tools
        return await self.client.chat.completions.create(**params)

    def _parse_response(
        self, resp: Any, request: ExecutionRequest, step: int
    ) -> ParsedTurn:
        if not resp.choices:
            raise RuntimeError(
                f"{self._label}: task {request.task_id}: step {step}: empty response"
            )

        choice = resp.choices[0]
        message = choice.message

        text = message.content if message.content else None
        is_complete = choice.finish_reason != "tool_calls"

        tool_calls: list[ToolInvocation] = []
        if not is_complete:
            for tool_call in message.tool_calls or []:
                tool_calls.append(
                    ToolInvocation(
                        id=tool_call.id,
                        name=tool_call.function.name,
                        raw_arguments=tool_call.function.arguments,
                    )
                )

        assistant_message = None if is_complete else message.model_dump()
        return ParsedTurn(
            text=text,
            tool_calls=tool_calls,
            is_complete=is_complete,
            assistant_message=assistant_message,
        )

    def _parse_arguments(self, raw_arguments: Any) -> dict[str, Any]:
        return json.loads(raw_arguments)

    def _format_tool_result(
        self, call: ToolInvocation, content: str, is_error: bool
    ) -> dict[str, Any]:
        return {
            "role": "tool",
            "tool_call_id": call.id,
            "content": content,
        }

    def _append_tool_results(self, messages: list[Any], tool_results: list[Any]) -> None:
        messages.extend(tool_results)
